packages/plugin/vite/src/ViteConfig.ts

```diff
@@ -0,0 +1,122 @@
+
+const d = debug('electron-forge:plugin:vite:viteconfig');
+
+/**
+ * If LoadResult is null, it means that Vite allows run without config file.
+ */
+export type LoadResult = Awaited<ReturnType<typeof loadConfigFromFile>>;
+
+export default class ViteConfigGenerator {
+  static resolveInlineConfig(CLIOptions: Record<string, any> = {}): InlineConfig {
+    // TODO: enhance
+    // https://github.com/vitejs/vite/blob/v4.0.1/packages/vite/src/node/cli.ts#L123-L132
+    const { root, base, mode, config: configFile, logLevel, clearScreen, force, ...server } = CLIOptions;
+    return {
+      root,
+      base,
+      mode,
+      configFile,
+      logLevel,
+      clearScreen,
+      optimizeDeps: { force },
+      server,
+    };
+  }
+
+  private isProd: boolean;
+
+  private pluginConfig: VitePluginConfig;
+
+  private projectDir!: string;
+
+  private baseDir!: string;
+
+  // Renderer config
+  loadResult: LoadResult;
+
+  // Vite's command config.
+  inlineConfig: InlineConfig;
+
+  constructor(pluginConfig: VitePluginConfig, projectDir: string, isProd: boolean, loadResult: LoadResult) {
+    this.pluginConfig = pluginConfig;
+    this.projectDir = projectDir;
+    this.baseDir = path.join(projectDir, '.vite');
+    this.isProd = isProd;
+    this.loadResult = loadResult;
+    this.inlineConfig = ViteConfigGenerator.resolveInlineConfig(this.pluginConfig.CLIOptions);
+
+    d('Config mode:', this.mode);
+  }
+
+  async resolveConfig(config: string, configEnv: Partial<ConfigEnv> = {}) {
+    configEnv.command ??= 'build'; // should always build.
+    configEnv.mode ??= this.mode;
+    return loadConfigFromFile(configEnv as ConfigEnv, config);
+  }
+
+  get mode(): string {
+    return this.inlineConfig.mode ?? (this.isProd ? 'production' : 'development');
+  }
+
+  getDefines(): Record<string, string> {
+    const port = this.loadResult?.config.server?.port ?? 5173;
+    return { VITE_DEV_SERVER_URL: this.isProd ? (undefined as any) : `'http://localhost:${port}'` };
+  }
+
+  getBuildConfig(watch = false): Promise<UserConfig>[] {
+    if (!Array.isArray(this.pluginConfig.build)) {
+      throw new Error('"config.build" must be an Array');
+    }
+
+    return this.pluginConfig.build
+      .filter(({ entry, config }) => entry || config)
+      .map(async ({ entry, config }) => {
+        const defaultConfig: UserConfig = {
+          // Ensure that each build config loads the .env file correctly.
+          mode: this.mode,
+          build: {
+            lib: entry
+              ? {
+                  entry,
+                  // Electron can only support cjs.
+                  formats: ['cjs'],
+                  fileName: () => '[name].js',
+                }
+              : undefined,
+            // Prevent multiple builds from interfering with each other.
+            emptyOutDir: false,
+            // 🚧 Multiple builds may conflict.
+            outDir: path.join(this.baseDir, 'build'),
+            watch: watch ? {} : undefined,
```

Comment:
question: does the main process get watched and reload on every change here?

Reply:
The Main process file is rebuilt immediately after the change, but it doesn't restart Electron APP because I don't know how it starts.
If there was a `restart` method available to me, then it would work with `watch`(rebuilt) to provide a better development experience.

Reply:
`=。=` Many users may not know that you can restart APP by typing `rs` on the command line, I even saw a trainer keep ending and restarting the command line every time he changed the main process code while giving a lecture to his students.

Vite can work with `watch` mode to restart APP automatically, not sure if I need this feature built-in?

**e.g.**

```js
// forge.config.js
module.exports = {
  plugins: [
    {
      name: '@electron-forge/plugin-vite',
      config: {
        build: [
          {
            entry: 'src/main.js',
            config: 'vite.main.config.mjs',
            autoRestart: true, // Here
          },
        ],
      },
    },
  ],
};
```

Or we can write a separate vite plugin and publish it to npm?

**e.g.**

```js
import autoRestart from 'vite-plugin-forge-auto-restart'

// vite.main.config.mjs
export default {
  plugins: [autoRestart()],
}
```